src/ariel-os-macros/src/define_count_adjusted_sensor_enums.rs

```diff
@@ -66,6 +70,54 @@ pub fn define_count_adjusted_sensor_enums(_item: TokenStream) -> TokenStream {
     });
 
     let expanded = quote! {
+        /// Provides access to the sensor driver instance.
+        /// For driver implementors only.
+        pub trait SensorAccess: private::Sealed {
+            /// Returns the sensor driver instance that produced these samples.
+            /// For driver implementors only.
+            fn sensor(&self) -> &'static dyn Sensor;
+        }
+
+        /// Avoid external implementations of [`SensorAccess`].
+        mod private {
+            use super::Samples;
+            pub trait Sealed {}
+
+            impl Sealed for Samples {}
+        }
+
+        #[derive(Copy, Clone, Debug)]
+        struct SizedIterator<I: Iterator> {
+            iter: I,
+            size: usize,
+        }
+
+        impl<I: Iterator> SizedIterator<I> {
+            fn new(iter: I, size: usize) -> Self {
+                Self { iter, size }
+            }
+        }
+
+        impl<I: Iterator> Iterator for SizedIterator<I> {
+            type Item = I::Item;
+
+            fn next(&mut self) -> Option<Self::Item> {
+                self.iter.next()
+            }
+            fn size_hint(&self) -> (usize, Option<usize>) {
+                (self.size, Some(self.size))
+            }
+        }
+
+        impl<I: Iterator> ExactSizeIterator for SizedIterator<I> {}
+        impl<I: Iterator + core::iter::FusedIterator> core::iter::FusedIterator for SizedIterator<I> {}
+
+        /// Return all samples of a sensor, without filtering out opaque channels. For driver implementors only.
+        pub trait UnfilteredSamples {
```

Comment:
I think we could call it that: the other, filtered trait is called `Reading`, and this is in a way a private interface, in that there is no guarantee about provided to driver users.

```suggestion
        pub trait PrivateReading {
```